scapy/contrib/mac_control.py

```diff
@@ -0,0 +1,255 @@
+          (https://github.com/wireshark/wireshark/blob/master/epan/dissectors/packet-maccontrol.c)
+
+"""
+
+from scapy.compat import orb
+from scapy.data import ETHER_TYPES
+from scapy.fields import IntField, ByteField, ByteEnumField, ShortField, BitField
+from scapy.layers.dot11 import Packet
+from scapy.layers.l2 import Ether, Dot1Q, bind_layers
+
+MAC_CONTROL_ETHER_TYPE = 0x8808
+ETHER_TYPES['MAC_CONTROL'] = MAC_CONTROL_ETHER_TYPE
+
+
+class MACControl(Packet):
+    DEFAULT_DST_MAC = "01:80:c2:00:00:01"
+
+    OP_CODE_PAUSE = 0x0001
+    OP_CODE_GATE = 0x0002
+    OP_CODE_REPORT = 0x0003
+    OP_CODE_REGISTER_REQ = 0x0004
+    OP_CODE_REGISTER = 0x0005
+    OP_CODE_REGISTER_ACK = 0x0006
+    OP_CODE_CLASS_BASED_FLOW_CONTROL = 0x0101
+
+    OP_CODES = {
+        OP_CODE_PAUSE: 'pause',
+        OP_CODE_GATE: 'gate',
+        OP_CODE_REPORT: 'report',
+        OP_CODE_REGISTER_REQ: 'register req',
+        OP_CODE_REGISTER: 'register',
+        OP_CODE_REGISTER_ACK: 'register_ack',
+        OP_CODE_CLASS_BASED_FLOW_CONTROL: 'class based flow control'
+    }
+
+    '''
+    flags used by Register* messages
+    '''
+    FLAG_REGISTER = 0x01
+    FLAG_DEREGISTER = 0x02
+    FLAG_ACK = 0x03
+    FLAG_NACK = 0x04
+
+    REGISTER_FLAGS = {
+        FLAG_REGISTER: 'register',
+        FLAG_DEREGISTER: 'deregister',
+        FLAG_ACK: 'ack',
+        FLAG_NACK: 'nack'
+    }
+
+    def guess_payload_class(self, payload):
+
+        try:
+            op_code = (orb(payload[0]) << 8) + orb(payload[1])
+            return MAC_CTRL_CLASSES[op_code]
+        except KeyError:
+            pass
+
+        return Packet.guess_payload_class(self, payload)
+
+    def _get_underlayers_size(self):
+        """
+        get the total size of all under layers
+        :return: number of bytes
+        """
+
+        under_layer = self.underlayer
+
+        under_layers_size = 0
+
+        while under_layer and isinstance(under_layer, Dot1Q):
+            under_layers_size += 4
+            under_layer = under_layer.underlayer
+
+        if under_layer and isinstance(under_layer, Ether):
+            # ether header len + FCS len
+            under_layers_size += 14 + 4
+
+        return under_layers_size
+
+    def post_build(self, pkt, pay):
+        """
+        add padding to the frame if required.
+
+        note that padding is only added if pay is None/empty. this allows us to add
+        any payload after the MACControl* PDU if needed (piggybacking).
+        """
+
+        if not pay:
+            under_layers_size = self._get_underlayers_size()
+
+            frame_size = (len(pkt) + under_layers_size)
+
+            if frame_size < 64:
+                return pkt + b'\x00' * (64 - frame_size)
+
+        return pkt + pay
+
+
+class MACControlPause(MACControl):
+    fields_desc = [
+        ShortField("_op_code", MACControl.OP_CODE_PAUSE),
+        ShortField("pause_time", 0),
+    ]
+
+    def get_pause_time_10_mbit(self):
```

Comment:
Looks redundant. Can't you use a single method that accept the data rate as an argument?

Reply:
If functional decomposition goes mad... hope aa4ff56 fixes this issue in a reasonable way.